DHDPS-397: CanDIG-API Schema

OrdiNeu opened the pull request:
# Ticket(s)

- [DHDPS-397](https://tfri.atlassian.net/browse/DHDPS-397)

# Description
This introduces an OpenAPI spec that we can start to implement with this microservice. Still WIP while I hash out the return values.

## Types of Change(s)

- [ ] 🪲 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [] 💥 Breaking change (fix or feature that would cause existing functionality to change)

<!--- If a breaking change, describe the impact the PR will have on the services selected above --->

## Has it been tested for:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] Prettier linter doesn't return errors
- [ ] Locally tested
  - [ ] using stable release
  - [ ] using develop branches
- [ ] Dev server tested
- [ ] Production tested when merging into stable/production branch

OrdiNeu committed:
[wip]

OrdiNeu committed:
[wip] Schema errors fixed

OrdiNeu committed:
[wip] Complete up to the request/response bodies

kcranston commented on schema.yml:
```diff
@@ -0,0 +1,831 @@
+  
+  /datasets:
+    get:
+      summary: List datasets. With parameters, searches through all datasets by one or more parameters.
```
not yet sure we need a search on this endpoint, or if beacon will be able to handle all of the use cases. Perhaps remove until we know we need it?

OrdiNeu replied:
Would that also be true of the persons search?

kcranston commented on schema.yml:
```diff
@@ -0,0 +1,831 @@
+        datasets instead. The granularity of the response is dependent on the authorized level
+        of access of the user
+      operationId: dataset_operations.search
+      parameters:
```
if we do incorporate search here, we definitely don't want a hardcoded list of parameters. Want a /datasets/parameters endpoint that lists the available search terms, e.g. beacon's filtering terms endpoint https://github.com/ga4gh-beacon/beacon-v2/blob/c6558bf2e6494df3905f7b2df66e903dfe509500/models/json/beacon-v2-default-model/endpoints.json#L151

OrdiNeu replied:
Done, I've copied over most of the Beacon stuff into beacon-schema.yaml which should let us avoid using a hardcoded parameters list

kcranston commented on schema.yml:
```diff
@@ -0,0 +1,831 @@
+            application/json:
+              schema:
+                oneOf:
+                  - $ref: '#/components/schemas/DatasetAggregateBody'
```
A dataset might contains lots of different types of data, so I think we will want more response schemas, or more endpoints within datasets to allow users to say "only clinical data" or "clinical + variants" or "images only". Maybe something like `datasets/{id}/clinical` or `datasets/{id}/variants`?

kcranston commented on schema.yml:
```diff
@@ -0,0 +1,831 @@
+          $ref: "#/components/responses/5xxServerError"
+      tags:
+        - datasets
+    post:
```
this probably also needs subsections for uploading different types of data to a dataset (clinical, genomic, images)

kcranston commented on schema.yml:
```diff
@@ -0,0 +1,831 @@
+        of access of the user, but usually returns only summary-level information. To obtain more
+        information about a person, use the /persons/{id} endpoint.
+      operationId: person_operations.search
+      parameters:
```
same as datasets, we may not need search here if we can do this all via beacon. But if we do, we don't want to hardcode all of the search parameters

kcranston commented on schema.yml:
```diff
@@ -0,0 +1,831 @@
+      tags:
+        - persons
+    post:
+      summary: Insert a new person into the database.
```
need to be able to add more info to an existing person, or modify some info

kcranston commented on schema.yml:
```diff
@@ -0,0 +1,831 @@
+            required:
+              - meta
+              - query
+  parameters:
```
to make this more extensible and future-proof, let's not hard code parameters into the schema, but let the service return these dynamically via a 'search_parameters' or 'filtering_parameters' endpoint

OrdiNeu replied:
Done, I've copied over most of the Beacon stuff into beacon-schema.yaml which should let us avoid using a hardcoded parameters list

kcranston reviewed (changes requested):
It might be easier to separate the beacon schema and the non-beacon schema into two smaller separate yml docs.

OrdiNeu committed:
Rework after discussions yesterday

OrdiNeu committed:
[wip] Remove unused fields

OrdiNeu committed:
[wip] response details

OrdiNeu committed:
Split the docker schema into its own .yml

OrdiNeu committed:
Add filtering terms

OrdiNeu committed:
Just use $refs to the actual BeaconAPI spec where possible

OrdiNeu committed:
Finish unrolling beacon schema, also misnamed docker-schema->beacon-schema

OrdiNeu committed:
Add missing (required) Beacon endpoints

SonQBChau commented on beacon-schema.yml:
```diff
@@ -0,0 +1,1304 @@
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/requestBodies/BeaconRequestBody'
```
Does the 200 response have the same ref with the request body?#/components/requestBodies/BeaconRequestBody

SonQBChau commented on beacon-schema.yml:
```diff
@@ -0,0 +1,1304 @@
+        responseSummary:
+          description: >-
+            Response summary, including Boolean and optionally results count.
+            $ref: '#/components/schemas/beaconCountResponseSection'
```
couldn't find this ref

SonQBChau commented on beacon-schema.yml:
```diff
@@ -0,0 +1,1304 @@
+          $ref: '#/components/schemas/PageSize'
+        skip:
+          $ref: '#/components/schemas/Skip'
+    RequestParameters:
```
need to define this object to implement

SonQBChau commented on beacon-schema.yml:
```diff
@@ -0,0 +1,1304 @@
+              schema:
+                $ref: '#/components/responses/BeaconErrorResponse'
+    post:
+      description: Get the list of filtering terms that could be used with datasets.
```
Is this the same with GET? Do we need both?

SonQBChau commented on schema.yml:
```diff
@@ -0,0 +1,646 @@
+        schema:
+          title: Dataset Id
+          type: string
+      responses:
```
This need a requestBody

SonQBChau commented on schema.yml:
```diff
@@ -0,0 +1,646 @@
+        schema:
+          title: Dataset Id
+          type: string
+      responses:
```
same, need a requestBody

SonQBChau commented on schema.yml:
```diff
@@ -0,0 +1,646 @@
+        schema:
+          title: Person Id
+          type: string
+      responses:
```
need requestBody

SonQBChau commented on schema.yml:
```diff
@@ -0,0 +1,646 @@
+        schema:
+          title: Person Id
+          type: string
+      responses:
```
need requestBody

SonQBChau commented on schema.yml:
```diff
@@ -0,0 +1,646 @@
+      description: |
+        Create a dataset corresponding to the given ID. If the dataset already exists, it will be
+        replaced.
+      operationId: dataset_operations.post_by_id
```
should rename to dataset_operations.put_by_id?

SonQBChau commented on schema.yml:
```diff
@@ -0,0 +1,646 @@
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/PersonAggregateBody'
```
We can't return the person's data after it's been deleted

SonQBChau replied:
We can't return the person's data after it's been deleted

SonQBChau reviewed (commented):
Looks good overall. I found a few minor issues. Thanks for taking the first step in designing v3!

SonQBChau commented:
Also, there is a small issue when I tried it. There are tabs in the YAML files, and the compiler didn't like it. Can you replace the tabs with spaces?

SonQBChau commented:
I found a new issue: the api `/datasets/{id}` returns `DatasetAggregateBody`, which is an array. Since it fetch a single item, shouldn't it return a single object instead?

OrdiNeu committed:
Post->put

OrdiNeu committed:
ADd single dataset responses

OrdiNeu committed:
Replace tabs with spaces

OrdiNeu commented:
> I found a new issue: the api `/datasets/{id}` returns `DatasetAggregateBody`, which is an array. Since it fetch a single item, shouldn't it return a single object instead?

Fixed in [5ece7c4](https://github.com/CanDIG/candig-api/pull/2/commits/5ece7c48728c3e7f7b221c73f00e930fab524ee9)

> Also, there is a small issue when I tried it. There are tabs in the YAML files, and the compiler didn't like it. Can you replace the tabs with spaces? 

Fixed in [9970826](https://github.com/CanDIG/candig-api/pull/2/commits/99708263eb6a3851eb2e7d44fc03dfd686178ff1)

OrdiNeu committed:
Respond to a PUT request with a single dataset's info, with no persons in the results

SonQBChau commented:
The `GET /datasets/{id}` still not working. I tracked down the reason and it caused by returns oneOf:
```
	- $ref: '#/components/schemas/DatasetSingleNoMetadataBody'
	- $ref: '#/components/schemas/DatasetSingleBody'
```
	
Because metadata is optional in DatasetSingleBody, it make {'id': 'Dataset_A', 'count': 6} ambiguous because matches both schemas. Type object is also missing in the schema

Here the full error:
```
Validation error: {'id': 'Dataset_A', 'count': 6} is valid under each of {'type': 'object', 'properties': {'id': {'type': 'string'}, 'results': {'type': 'array', 'items': {'type': 'array', 'items': {'type': 'object', 'properties': {'person_id': {'description': 'Any person linkage that needs to occur to uniquely identify Persons ought to be done prior to writing this table. This identifier can be the original id from the source data provided if it is an integer, otherwise it can be an autogenerated number. It is assumed that every person with a different unique identifier is in fact a different person and should be treated independently.', 'type': 'integer'}, 'gender_concept_id': {'description': 'Use the gender or sex value present in the data under the assumption that it is the biological sex at birth. If the source data captures gender identity it should be stored in the OBSERVATION table. Accepted gender concepts. Please refer to the THEMIS repository for detailed conventions on how to populate this field. This field is meant to capture the biological sex at birth of the Person. This field should not be used to study gender identity issues.', 'type': 'integer'}, 'year_of_birth': {'description': "For data sources with date of birth, the year should be extracted. If no year of birth is available all the person's data should be dropped from the CDM instance. For additional information on how to populate this field, please refer to the THEMIS repository. Compute age using year_of_birth.", 'type': 'integer'}, 'month_of_birth': {'description': 'For data sources that provide the precise date of birth, the month should be extracted and stored in this field.', 'type': 'integer'}, 'day_of_birth': {'description': 'For data sources that provide the precise date of birth, the day should be extracted and stored in this field.', 'type': 'integer'}, 'birth_datetime': {'description': 'This field is not required but highly encouraged. For data sources that provide the precise datetime of birth, that value should be stored in this field. For more information on how to populate this field, please refer to the THEMIS repository.', 'type': 'string'}, 'race_concept_id': {'description': 'This field captures race or ethnic background of the person.  Only use this field if you have information about race or ethnic background. The Vocabulary contains Concepts about the main races and ethnic backgrounds in a hierarchical system. Due to the imprecise nature of human races and ethnic backgrounds, this is not a perfect system. Mixed races are not supported. If a clear race or ethnic background cannot be established, use Concept_Id.', 'type': 'integer'}, 'ethnicity_concept_id': {'description': 'This field captures Ethnicity as defined by the Office of Management and Budget (OMB) of the US Government - it distinguishes only between "Hispanic" and "Not Hispanic". Races and ethnic backgrounds are not stored here. Only use this field if you have US-based data and a source of this information. Do not attempt to infer Ethnicity from the race or ethnic background of the Person.', 'type': 'integer'}, 'location_id': {'description': 'The location refers to the physical address of the person. This field should capture the last known location of the person.  Put the location_id from the LOCATION table here that represents the most granular location information for the person. For additional information on how to populate this field, please refer to the THEMIS repository.', 'type': 'integer'}, 'provider_id': {'description': 'The Provider refers to the last known primary care provider (General Practitioner).  Put the provider_id from the PROVIDER table of the last known general practitioner of the person. If there are multiple providers, it is up to the ETL to decide which to put here.', 'type': 'integer'}, 'care_site_id': {'description': 'The Care Site refers to where the Provider typically provides the primary care.', 'type': 'integer'}, 'person_source_value': {'description': 'Use this field to link back to persons in the source data. This is typically used for error checking of ETL logic. Some use cases require the ability to link back to persons in the source data. This field allows for the storing of the person value as it appears in the source. This field is not required but strongly recommended.', 'type': 'string'}, 'gender_source_value': {'description': 'This field is used to store the biological sex of the person from the source data. It is not intended for use in standard analytics but for reference only. Put the assigned sex at birth of the person as it appears in the source data.', 'type': 'string'}, 'gender_source_concept_id': {'description': 'Due to the small number of options, this tends to be zero. If the source data codes assigned sex at birth in a non-standard vocabulary, store the concept_id here.', 'type': 'integer'}, 'race_source_value': {'description': 'This field is used to store the race of the person from the source data. It is not intended for use in standard analytics but for reference only. Put the race of the person as it appears in the source data.', 'type': 'string'}, 'race_source_concept_id': {'description': 'Due to the small number of options, this tends to be zero. If the source data codes race in an OMOP supported vocabulary store the concept_id here.', 'type': 'integer'}, 'ethnicity_source_value': {'description': 'This field is used to store the ethnicity of the person from the source data. It is not intended for use in standard analytics but for reference only. If the person has an ethnicity other than the OMB standard of "Hispanic" or "Not Hispanic" store that value from the source data here.', 'type': 'string'}, 'ethnicity_source_concept_id': {'description': 'Due to the small number of options, this tends to be zero. If the source data codes ethnicity in an OMOP supported vocabulary, store the concept_id here.', 'type': 'integer'}}}}}, 'metadata': {'type': 'object', 'properties': {'cohort_definition_id': {'description': 'This is the identifier given to the cohort, usually by the ATLAS application', 'type': 'string'}, 'cohort_definition_name': {'description': 'A short description of the cohort', 'type': 'string'}, 'cohort_definition_description': {'description': 'A complete description of the cohort.', 'type': 'string'}, 'definition_type_concept_id': {'description': 'Type defining what kind of Cohort Definition the record represents and how the syntax may be executed.', 'type': 'integer'}, 'cohort_definition_syntax': {'description': 'Syntax or code to operationalize the Cohort Definition.', 'type': 'string'}, 'subject_concept_id': {'description': 'This field contains a Concept that represents the domain of the subjects that are members of the cohort (e.g., Person, Provider, Visit).', 'type': 'integer'}, 'cohort_initiation_date': {'description': 'A date to indicate when the Cohort was initiated in the COHORT table.', 'type': 'string'}}}}}, {'type': 'object', 'properties': {'id': {'type': 'string'}, 'count': {'type': 'integer'}}}
NonConformingResponseBody(status_code=500, detail='Response body does not conform to specification. {\'id\': \'Dataset_A\', \'count\': 6} is valid under each of {\'type\': \'object\', \'properties\': {\'id\': {\'type\': \'string\'}, \'results\': {\'type\': \'array\', \'items\': {\'type\': \'array\', \'items\': {\'type\': \'object\', \'properties\': {\'person_id\': {\'description\': \'Any person linkage that needs to occur to uniquely identify Persons ought to be done prior to writing this table. This identifier can be the original id from the source data provided if it is an integer, otherwise it can be an autogenerated number. It is assumed that every person with a different unique identifier is in fact a different person and should be treated independently.\', \'type\': \'integer\'}, \'gender_concept_id\': {\'description\': \'Use the gender or sex value present in the data under the assumption that it is the biological sex at birth. If the source data captures gender identity it should be stored in the OBSERVATION table. Accepted gender concepts. Please refer to the THEMIS repository for detailed conventions on how to populate this field. This field is meant to capture the biological sex at birth of the Person. This field should not be used to study gender identity issues.\', \'type\': \'integer\'}, \'year_of_birth\': {\'description\': "For data sources with date of birth, the year should be extracted. If no year of birth is available all the person\'s data should be dropped from the CDM instance. For additional information on how to populate this field, please refer to the THEMIS repository. Compute age using year_of_birth.", \'type\': \'integer\'}, \'month_of_birth\': {\'description\': \'For data sources that provide the precise date of birth, the month should be extracted and stored in this field.\', \'type\': \'integer\'}, \'day_of_birth\': {\'description\': \'For data sources that provide the precise date of birth, the day should be extracted and stored in this field.\', \'type\': \'integer\'}, \'birth_datetime\': {\'description\': \'This field is not required but highly encouraged. For data sources that provide the precise datetime of birth, that value should be stored in this field. For more information on how to populate this field, please refer to the THEMIS repository.\', \'type\': \'string\'}, \'race_concept_id\': {\'description\': \'This field captures race or ethnic background of the person.  Only use this field if you have information about race or ethnic background. The Vocabulary contains Concepts about the main races and ethnic backgrounds in a hierarchical system. Due to the imprecise nature of human races and ethnic backgrounds, this is not a perfect system. Mixed races are not supported. If a clear race or ethnic background cannot be established, use Concept_Id.\', \'type\': \'integer\'}, \'ethnicity_concept_id\': {\'description\': \'This field captures Ethnicity as defined by the Office of Management and Budget (OMB) of the US Government - it distinguishes only between "Hispanic" and "Not Hispanic". Races and ethnic backgrounds are not stored here. Only use this field if you have US-based data and a source of this information. Do not attempt to infer Ethnicity from the race or ethnic background of the Person.\', \'type\': \'integer\'}, \'location_id\': {\'description\': \'The location refers to the physical address of the person. This field should capture the last known location of the person.  Put the location_id from the LOCATION table here that represents the most granular location information for the person. For additional information on how to populate this field, please refer to the THEMIS repository.\', \'type\': \'integer\'}, \'provider_id\': {\'description\': \'The Provider refers to the last known primary care provider (General Practitioner).  Put the provider_id from the PROVIDER table of the last known general practitioner of the person. If there are multiple providers, it is up to the ETL to decide which to put here.\', \'type\': \'integer\'}, \'care_site_id\': {\'description\': \'The Care Site refers to where the Provider typically provides the primary care.\', \'type\': \'integer\'}, \'person_source_value\': {\'description\': \'Use this field to link back to persons in the source data. This is typically used for error checking of ETL logic. Some use cases require the ability to link back to persons in the source data. This field allows for the storing of the person value as it appears in the source. This field is not required but strongly recommended.\', \'type\': \'string\'}, \'gender_source_value\': {\'description\': \'This field is used to store the biological sex of the person from the source data. It is not intended for use in standard analytics but for reference only. Put the assigned sex at birth of the person as it appears in the source data.\', \'type\': \'string\'}, \'gender_source_concept_id\': {\'description\': \'Due to the small number of options, this tends to be zero. If the source data codes assigned sex at birth in a non-standard vocabulary, store the concept_id here.\', \'type\': \'integer\'}, \'race_source_value\': {\'description\': \'This field is used to store the race of the person from the source data. It is not intended for use in standard analytics but for reference only. Put the race of the person as it appears in the source data.\', \'type\': \'string\'}, \'race_source_concept_id\': {\'description\': \'Due to the small number of options, this tends to be zero. If the source data codes race in an OMOP supported vocabulary store the concept_id here.\', \'type\': \'integer\'}, \'ethnicity_source_value\': {\'description\': \'This field is used to store the ethnicity of the person from the source data. It is not intended for use in standard analytics but for reference only. If the person has an ethnicity other than the OMB standard of "Hispanic" or "Not Hispanic" store that value from the source data here.\', \'type\': \'string\'}, \'ethnicity_source_concept_id\': {\'description\': \'Due to the small number of options, this tends to be zero. If the source data codes ethnicity in an OMOP supported vocabulary, store the concept_id here.\', \'type\': \'integer\'}}}}}, \'metadata\': {\'type\': \'object\', \'properties\': {\'cohort_definition_id\': {\'description\': \'This is the identifier given to the cohort, usually by the ATLAS application\', \'type\': \'string\'}, \'cohort_definition_name\': {\'description\': \'A short description of the cohort\', \'type\': \'string\'}, \'cohort_definition_description\': {\'description\': \'A complete description of the cohort.\', \'type\': \'string\'}, \'definition_type_concept_id\': {\'description\': \'Type defining what kind of Cohort Definition the record represents and how the syntax may be executed.\', \'type\': \'integer\'}, \'cohort_definition_syntax\': {\'description\': \'Syntax or code to operationalize the Cohort Definition.\', \'type\': \'string\'}, \'subject_concept_id\': {\'description\': \'This field contains a Concept that represents the domain of the subjects that are members of the cohort (e.g., Person, Provider, Visit).\', \'type\': \'integer\'}, \'cohort_initiation_date\': {\'description\': \'A date to indicate when the Cohort was initiated in the COHORT table.\', \'type\': \'string\'}}}}}, {\'type\': \'object\', \'properties\': {\'id\': {\'type\': \'string\'}, \'count\': {\'type\': \'integer\'}}}')
```

SonQBChau commented:
There is a few issues still need fixing:
	- ~~`PUT /datasets/{id}` – missing requestBody~~
	- ~~`PATCH /datasets/{id}/metadata` – missing requestBody~~
	- ~~`PUT /datasets/{dataset_id}/persons/{id}` – missing requestBody~~
	- ~~`PATCH /datasets/{dataset_id}/persons/{id}` – missing requestBody~~
	- ~~`DELETE /datasets/{dataset_id}/persons/{id}` – should not return anything. It should return 204 or a message~~
	- ~~`GET /datasets/{id} – returns DatasetMetadataBody` which still references the OMOP cohort object~~
	- ~~Rename `metadata` to `dataset_info` in datasets object~~
	- ~~Rename `metadata` to `dataset_info` in /datasets/{id}/metadata. The dataset_info object also need to be updated (not cohort anymore)~~

OrdiNeu committed:
Rename metadata->dataset_info in response

OrdiNeu committed:
Return is ambiguous

OrdiNeu committed:
Rename Results in the response to Persons. Add requestBodies

SonQBChau committed:
Update schema.yml

- rename metadata to info to avoid confusion
- replace DatasetMetadataBody

SonQBChau committed:
add nullable and require FK to person

SonQBChau committed:
simplify DatasetStatsBody

SonQBChau committed:
Update schema.yml

- Move /datasets/statistics above /datasets/{id}
- Add IndividualDatasetStatsBody
- Format yml

SonQBChau reviewed (approved):
approve so we can merge and continue later
